### PR TITLE
feat: Add prettifier util function with unit tests

### PR DESCRIPTION
**Pull request by zzxthehappiest**
# Description

This PR is to import a previouly implemented [function](https://github.com/junhaoliao/yscope-log-viewer/blob/beta-release/src/Viewer/services/decoder/SimplePrettifier.js) as discussed in #220  to replace the `js-beautify` to format the object, tuple and array structures nested in the log message. This function will be used after the PR #223 is merged.

This PR also fixes a minor bug in the original [code](https://github.com/junhaoliao/yscope-log-viewer/blob/beta-release/src/Viewer/services/decoder/SimplePrettifier.js#L90), where we should check `\` rather than `'`. This PR also refactors the code to make can pass the lint because the original function is too long.

This PR also adds unit tests to cover the basic use cases and corner cases.

# Checklist

* [ ] The PR satisfies the [contribution guidelines][yscope-contrib-guidelines].
* [ ] This is a breaking change and that has been indicated in the PR title, OR this isn't a
  breaking change.
* [ ] Necessary docs have been updated, OR no docs need to be updated.

# Validation performed

Passed all unit tests.

[yscope-contrib-guidelines]: https://docs.yscope.com/dev-guide/contrib-guides-overview.html

### CODE REVIEW

**Comment by junhaoliao** [comment]
Big thanks for porting this over! Let's see if @kirkrodrigues is available to review this.

**Comment by kirkrodrigues** [comment]
Sorry for the delay. Will get to this during the week.

**Comment by kirkrodrigues** [comment]
> Sorry for the delay. Will get to this during the week.

Sorry, need a little more time to finish writing up my thoughts. Hopefully in the next 1-2 days.

File: src/utils/prettifier.ts [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+        }
+    }
+
+    if (state.copyBeginOffset < uglyString.length) {
+        state.result += uglyString.substring(state.copyBeginOffset);
+    }
+
+    return [state.isPrettifiedStringDifferent,
+        state.result];
+};
```
Comment (kirkrodrigues) [inline src/utils/prettifier.ts:290]: Thanks for implementing this. I think your implementation is more clear than the original one, but I think the original algorithm was a little confusing. So I wonder if you could try something a little different.

I'm roughly thinking that this method could be written like this (wrote it in Python for simplicity):

```python
# TODO: Set state to whatever we need

i = 0
is_escaped = False
while i < len(ugly):
    c = ugly[i]

    if is_escaped:
        is_escaped = False
        i += 1
    elif '\\' == c:
        is_escaped = True
        i += 1
    elif '"' == c or '\'' == c:
        i = find_next_unescaped_char(ugly, i + 1, '\\', c)
    elif '{' == c or '(' == c:
        i = handle_open_brace_or_paren(ugly, i, state)
    elif '}' == c or ')' == c:
        i = handle_close_brace_or_paren(ugly, i, state)
    elif '[' == c:
        i = handle_open_bracket(ugly, i, state)
    elif ']' == c:
        i = handle_close_bracket(ugly, i, state)
    elif ',' == c:
        i = handle_comma(ugly, i, state)
    else:
        i += 1

# TODO: Copy remaining characters
```

In particular, could we change the implementation as follows:

* Each function will return the index of the next character to consider.
* When we see a `"`, could we simply search for the next unescaped `"`? For example, you could write a function called `findNextUnescapedChar(uglyString, currentIdx, char, escapeChar)` that returns the index of the next unescaped character.
* Instead of tracking "array-indent-levels", could we maintain a stack that tracks what character will close the current region? For example, once we enter a braced (`{}`) region, we should push `}` onto the stack, indicating that we're expecting a `}` to close the current region.
  * This would also allow us to validate that whatever character we see actually closes the current region (otherwise, `{)` is a treated as valid region).
* After we encounter a `{`, `(`, `[`, or `,` (when it's inside a region), can we ignore all whitespace up to the next non-whitespace character? That way a string like "[    a, b, c]" would get formatted as "[a, b, c]".
* It would also be good if we could handle dangling commas without adding whitespace. For example, "[a, b, c, ]" should get formatted as "[a, b, c,]".

Do you think that makes sense?

File: test/utils/prettifier.test.ts
Comment (kirkrodrigues) [inline test/utils/prettifier.test.ts:1]: Thanks for implementing the tests. I'll review them in more detail once we finish the implementation.

File: src/utils/prettifier.ts
Comment (kirkrodrigues) [inline src/utils/prettifier.ts:1]: Let's wrap comment lines at 100 characters rather than 80 that we used in older code.

**Review by kirkrodrigues — CHANGES_REQUESTED** [review]

File: src/utils/prettifier.ts (hunk 1 above)
Comment (zzxthehappiest) [reply src/utils/prettifier.ts:290]: It is an amazing catch for this:

> This would also allow us to validate that whatever character we see actually closes the current region (otherwise, `{)` is a treated as valid region
